common/ensemble_building/ensemble_selection.py

```diff
@@ -0,0 +1,210 @@
+from collections import Counter
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+
+from sklearn.pipeline import Pipeline
+
+from .abstract_ensemble import AbstractEnsemble
+
+
+class EnsembleSelection(AbstractEnsemble):
+    def __init__(
+        self,
+        ensemble_size: int,
+        loss_fn: Callable[..., float],
+        loss_fn_args: Dict[str, Any],
+        random_state: np.random.RandomState,
+    ) -> None:
+        self.ensemble_size = ensemble_size
+        self.loss_fn: Optional[Callable[..., float]] = loss_fn
+        self.loss_fn_args = loss_fn_args
+        self.random_state = random_state
+
+    def __getstate__(self) -> Dict[str, Any]:
+        # Cannot serialize a metric if
+        # it is user defined.
+        # That is, if doing pickle dump
+        # the metric won't be the same as the
+        # one in __main__. we don't use the metric
+        # in the EnsembleSelection so this should
+        # be fine
+        self.loss_fn = None
+        self.loss_fn_args = {}
+        return self.__dict__
+
+    def fit(
+        self,
+        predictions: List[np.ndarray],
+        labels: np.ndarray,
+        identifiers: List[Tuple[int, int, float]],
+    ) -> AbstractEnsemble:
+        self.ensemble_size = int(self.ensemble_size)
+        if self.ensemble_size < 1:
+            raise ValueError("Ensemble size cannot be less than one!")
+
+        self._fit(predictions, labels)
+        self._calculate_weights()
+        self.identifiers_ = identifiers
+        return self
+
+    def _fit(
+        self,
+        predictions: List[np.ndarray],
+        labels: np.ndarray,
+    ) -> AbstractEnsemble:
+        """Fast version of Rich Caruana's ensemble selection method."""
+
+        if self.loss_fn is None:
+            raise ValueError(
+                "A function with signature Callable[[labels, predictions, ...], float]"
+                "is needed to perform ensemble selection."
+            )
+
+        self.num_input_models_ = len(predictions)
+
+        ensemble: List[np.ndarray] = []
+        trajectory: List[float] = []
+        order: List[int] = []
+
+        ensemble_size = self.ensemble_size
+
+        weighted_ensemble_prediction = np.zeros(
+            predictions[0].shape,
+            dtype=np.float64,
+        )
+        fant_ensemble_prediction = np.zeros(
+            weighted_ensemble_prediction.shape,
+            dtype=np.float64,
+        )
+        for _ in range(ensemble_size):
+            losses = np.zeros(
+                (len(predictions)),
+                dtype=np.float64,
+            )
+            s = len(ensemble)
+            if s > 0:
+                np.add(
```

Comment:
This part consumes the same amount as the following:
```
weighted_ensemble_prediction += ensemble[-1]
```
It is shorter and understandable.